# Day 6: Wait for it

https://adventofcode.com/2023/day/6

Elves are small and quick; their toy boat races go for millimeters and last for milliseconds. I have a tiny boat, I push a button to charge it, release the button and watch it go. For every millisecond I hold the button, the boat goes one millimeter per second faster. I'm given the duration of the race and the distance record, which I need to beat. How many ways can I do it? I need the product of those. 


In [1]:
sample_input = '''Time:      7  15   30
Distance:  9  40  200
'''

In [2]:
def convert_to_integers(list_of_strings):
    list_of_integers = []
    for entry in list_of_strings:
        list_of_integers.append(int(entry))
    return list_of_integers

def read_input(the_input):
    line = the_input.split('\n')
    times = convert_to_integers(line[0][line[0].find(":")+1:].split())
    distances = convert_to_integers(line[1][line[1].find(":")+1:].split())
    return times, distances

In [3]:
times, distances = read_input(sample_input)
print(times, distances)

[7, 15, 30] [9, 40, 200]


In [4]:
def calculate_distance(charge_time, race_duration):
    # velocity = charge_time
    # travel_time = race_duration - charge_time
    # distance_traveled = velocity * travel_time
    distance_traveled = charge_time * (race_duration - charge_time)
    return distance_traveled

calculate_distance(4,7)

12

In [5]:
def solve_puzzle(the_input, verbose=True):
    times, distances = read_input(the_input)
    for race_number in range(len(times)):
        race_duration = times[race_number]
        distance_record = distances[race_number]
        count_the_ways = 0
        print(f'Race number: {race_number} | Race duration: {race_duration}, Distance record: {distance_record}')
        for charge_time in range(1, race_duration):
            distance_traveled = calculate_distance(charge_time, race_duration)
            beat_the_record = (distance_traveled > distance_record)
            if beat_the_record:
                count_the_ways += 1
                if verbose: print(f'   {charge_time}  ==>  {distance_traveled}  {beat_the_record}  {count_the_ways}')
        print(f'{count_the_ways} ways')
    return count_the_ways

solve_puzzle(sample_input)

Race number: 0 | Race duration: 7, Distance record: 9
   2  ==>  10  True  1
   3  ==>  12  True  2
   4  ==>  12  True  3
   5  ==>  10  True  4
4 ways
Race number: 1 | Race duration: 15, Distance record: 40
   4  ==>  44  True  1
   5  ==>  50  True  2
   6  ==>  54  True  3
   7  ==>  56  True  4
   8  ==>  56  True  5
   9  ==>  54  True  6
   10  ==>  50  True  7
   11  ==>  44  True  8
8 ways
Race number: 2 | Race duration: 30, Distance record: 200
   11  ==>  209  True  1
   12  ==>  216  True  2
   13  ==>  221  True  3
   14  ==>  224  True  4
   15  ==>  225  True  5
   16  ==>  224  True  6
   17  ==>  221  True  7
   18  ==>  216  True  8
   19  ==>  209  True  9
9 ways


9

In [6]:
4 * 8 * 9

288

In [7]:
puzzle_input = '''Time:        58     99     64     69
Distance:   478   2232   1019   1071
'''

In [8]:
solve_puzzle(puzzle_input, verbose=False)

Race number: 0 | Race duration: 58, Distance record: 478
39 ways
Race number: 1 | Race duration: 99, Distance record: 2232
30 ways
Race number: 2 | Race duration: 64, Distance record: 1019
5 ways
Race number: 3 | Race duration: 69, Distance record: 1071
22 ways


22

In [9]:
39*30*5*22

128700

128700 is correct!

# Part 2

Actually... each line is all one big number.

In [10]:
sample_input = '''Time:      71530
Distance:  940200'''

In [11]:
solve_puzzle(sample_input, verbose=False)

Race number: 0 | Race duration: 71530, Distance record: 940200
71503 ways


71503

In [12]:
puzzle_input = '''Time:        58996469
Distance:   478223210191071
'''

That's a big number. We might want to calculate the charge time to meet the record, and the charge time for the maximum distance, and the number of solutions should be twice that difference. Approximately. I could struggle with off-by-one errors or I could just let it run. Let's try just running it.

In [13]:
solve_puzzle(puzzle_input, verbose=False)

Race number: 0 | Race duration: 58996469, Distance record: 478223210191071
39594072 ways


39594072

39594072 is correct. For a one-time thing, it was faster to wait a minute or so than to rewrite the code. I'm 2 days behind as it is.